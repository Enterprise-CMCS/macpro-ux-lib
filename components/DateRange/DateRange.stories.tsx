import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import { DateRange } from "./DateRange";

export default {
  title: "COMPONENTS/DateRange",
  component: DateRange,
  argTypes: {
    defaultStartDate: {
      description:
        "The start date picker input will set this value by default if it is a valid date. The date should be in the format mm/dd/yyyy.",
      control: { type: "text" },
    },
    startDate: {
      description:
        "The value given to the start date input. The date should be in the format mm/dd/yyyy",
      control: { type: "text" },
    },
    startInputId: {
      description: "A unique identifier for the start date input.",
    },
    startInputName: {
      description: "Name of the start date input field.",
    },
    startLabel: {
      description: "The label of the start date input.",
    },
    defaultEndDate: {
      description:
        "The end date picker input will set this value by default if it is a valid date. The date should be in the format mm/dd/yyyy.",
      control: { type: "text" },
    },
    endDate: {
      description:
        "The value given to the end date input. The date should be in the format mm/dd/yyyy.",
      control: { type: "text" },
    },
    endInputId: {
      description: "A unique identifier for the end date input.",
    },

    endInputName: {
      description: "Name of the end date input field.",
    },
    endLabel: {
      description: "The label of the end date input.",
    },
    disabled: {
      description:
        "Controls whether or not the date range pickers are disabled to the user.",
    },
    hint: {
      description:
        "Boolean that shows or hide the date format hint for both inputs, in the format mm/dd/yyyy.",
    },
  },
  args: {
    startInputId: "id-1",
    startInputName: "name-1",
    startLabel: "Start Date",
    endInputId: "id-2",
    endInputName: "name-2",
    endLabel: "End Date",
  },
  parameters: {
    docs: {
      description: {
        component:
          "A date range picker helps users select a range between two dates. When users need to schedule or record an event and benefit from the context of a calendar. When knowing the day of the week helps users choose a specific date.",
      },
    },
  },
} as ComponentMeta<typeof DateRange>;

const Template: ComponentStory<typeof DateRange> = ({ ...rest }) => (
  <DateRange {...rest} />
);

export const PrimaryDateRange = Template.bind({});

export const DateRangeWithDefaults = Template.bind({});
DateRangeWithDefaults.args = {
  defaultStartDate: "10/10/2022",
  defaultEndDate: "10/15/2022",
};

DateRangeWithDefaults.parameters = {
  docs: {
    description: {
      story:
        "A start and end date can be provided to help users selected a certain start or end range by default",
    },
  },
};
