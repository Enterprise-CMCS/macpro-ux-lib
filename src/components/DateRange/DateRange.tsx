import React, { useState } from "react";
import { Datefield } from "components/Datefield/Datefield";

type IntrinsicElements = JSX.IntrinsicElements["div"];
export interface Props extends IntrinsicElements {
  // id: string;
  // name: string;
  startLabel: string;
  endLabel: string;
  hint?: boolean;
  disabled?: boolean;
  minDate?: string;
  maxDate?: string;
  // value?: string;
}

/**
 * DateRange Component
 * @param {string}  id                     A unique identifier for the input.
 * @param {string}  name                   Name of the input field.
 * @param {string}  label                  Field label.
 * @param {boolean} [hint]                 Boolean that shows or hide the date format hint, in the format mm/dd/yyyy.
 * @param {string}  [disabled]             Controls whether or not the date picker is disabled to the user.
 * @param {string}  [minDate]              The date picker will not allow a date selection before this date. The date should be in the format mm/dd/yyyy
 * @param {string}  [maxDate]              The date picker will not allow a date selection after this date. The date should be in the format mm/dd/yyyy.
 * @param {string}  [value]                Value of the input element.
 */

export const DateRange: React.FC<Props> = ({
  hint = true,
  minDate,
  maxDate,
  disabled,
  startLabel,
  endLabel,
  ...rest
}) => {
  const [currentMinDate, setMinDate] = useState(minDate);
  const [currentMaxDate, setMaxDate] = useState(maxDate);

  return (
    <div {...rest}>
      <Datefield
        disabled={disabled}
        name="date-range-1"
        id=""
        hint={hint}
        maxDate={currentMaxDate}
        label={startLabel}
        value={currentMinDate}
      />
      <Datefield
        value={currentMaxDate}
        disabled={disabled}
        name="date-range-1"
        id=""
        label={endLabel}
        hint={hint}
        minDate={currentMinDate}
      />
    </div>
  );
};
