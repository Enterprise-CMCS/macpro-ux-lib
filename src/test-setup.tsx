import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
