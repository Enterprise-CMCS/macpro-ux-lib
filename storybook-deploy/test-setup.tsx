import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };

/**
 * Clean up attributes from the rendered component by removing the generated ids.
 *
 * Before:
 *   for="input-type-text-123456"
 *   id="input-type-text-123456"
 *
 * After:
 *   for="input-type-text"
 *   id="input-type-text"
 *
 * This is useful during snapshot testing. Snapshot comparisons will fail without addressing these items.
 *
 * @param {HTMLElement} container The component to clean up.
 * @param {prefixes}      string[]  The reference substrings to leave alone.
 */
