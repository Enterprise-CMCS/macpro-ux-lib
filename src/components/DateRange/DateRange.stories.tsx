import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import { DateRange } from "./DateRange";

export default {
  title: "COMPONENTS/DateRange",
  component: DateRange,
  argTypes: {
    minDate: {
      description:
        "The date picker will not allow a date selection before this date. The date should be in the format mm/dd/yyyy.",
      control: { type: "text" },
    },
    maxDate: {
      description:
        "The date picker will not allow a date selection after this date. The date should be in the format mm/dd/yyyy.",
      control: { type: "text" },
    },
    startDate: {
      description:
        "The value given to the start date input. The date should be in the format mm/dd/yyyy",
      control: { type: "text" },
    },
    startInputId: {
      description: "A unique identifier for the start date input.",
    },
    startInputName: {
      description: "Name of the start date input field.",
    },
    startLabel: {
      description: "The label of the start date input.",
    },

    endDate: {
      description:
        "The value given to the end date input. The date should be in the format mm/dd/yyyy.",
      control: { type: "text" },
    },
    endInputId: {
      description: "A unique identifier for the end date input.",
    },

    endInputName: {
      description: "Name of the end date input field.",
    },
    endLabel: {
      description: "The label of the end date input.",
    },
    disabled: {
      description:
        "Controls whether or not the date range pickers are disabled to the user.",
    },
    hint: {
      description:
        "Boolean that shows or hide the date format hint for both inputs, in the format mm/dd/yyyy.",
    },
    required: {
      description:
        "The date picker component will be required in terms of native form validation.",
    },
  },
  args: {
    startInputId: "id-1",
    startInputName: "name-1",
    startLabel: "Start Date",
    endInputId: "id-2",
    endInputName: "name-2",
    endLabel: "End Date",
  },
  parameters: {
    docs: {
      description: {
        component:
          "A date range picker helps users select a range between two dates. When users need to schedule or record an event and benefit from the context of a calendar. When knowing the day of the week helps users choose a specific date.",
      },
    },
  },
} as ComponentMeta<typeof DateRange>;

const Template: ComponentStory<typeof DateRange> = ({ ...rest }) => (
  <DateRange {...rest} />
);

export const PrimaryDateRange = Template.bind({});

export const DateRangeWithDefaults = Template.bind({});
DateRangeWithDefaults.args = {
  startDate: "10/10/2022",
  endDate: "10/14/2022",
};
