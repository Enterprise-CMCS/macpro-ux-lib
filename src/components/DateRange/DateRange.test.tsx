import React from "react";
import { DateRange } from "./DateRange";
import { fireEvent, screen, render } from "../../test-setup";

describe("Tests for the DateRange component.", () => {
  it("Should render without hint", () => {
    render(
      <DateRange
        data-testid="DateRange"
        startInputId="test-1"
        endInputId="test-1"
        endInputName="test-1"
        startInputName="test-1"
        startLabel="test-1"
        endLabel="test-1"
        startDate="09/10/2022"
        endDate="09/12/2022"
        hint={false}
      />
    );
    const DatefieldButton1 = screen.queryAllByTestId("calendar-button")[0];
    const DatefieldButton2 = screen.queryAllByTestId("calendar-button")[1];
    fireEvent.click(DatefieldButton1);
    const randomDayButton = screen.getAllByRole("button")[6];
    fireEvent.click(randomDayButton);
    fireEvent.click(DatefieldButton2);
    const randomDayButton2 = screen.getAllByRole("button")[6];
    fireEvent.click(randomDayButton2);

    const DatefieldComp = screen.getByTestId("DateRange");
    expect(DatefieldComp).toBeInTheDocument();
  });

  it("Should render with hint", () => {
    render(
      <DateRange
        data-testid="DateRange"
        startInputId="test-2"
        endInputId="test-2"
        endInputName="test-2"
        startInputName="test-2"
        startLabel="test-2"
        endLabel="test-2"
        hint={true}
      />
    );
    const DatefieldButton1 = screen.queryAllByTestId("calendar-button")[0];
    const DatefieldButton2 = screen.queryAllByTestId("calendar-button")[1];
    fireEvent.click(DatefieldButton2);
    fireEvent.click(DatefieldButton1);
    fireEvent.click(DatefieldButton2);
    fireEvent.click(DatefieldButton1);

    const DatefieldComp = screen.getByTestId("DateRange");
    expect(DatefieldComp).toBeInTheDocument();
  });

  it("Should render with a start date", () => {
    render(
      <DateRange
        data-testid="DateRange"
        startInputId="test-6"
        endInputId="test-6"
        endInputName="test-6"
        startInputName="test-6"
        startLabel="test-6"
        endLabel="test-6"
        startDate="09/10/2022"
        hint={true}
      />
    );
    const DatefieldButton2 = screen.queryAllByTestId("calendar-button")[1];
    fireEvent.click(DatefieldButton2);

    const DatefieldComp = screen.getByTestId("DateRange");
    expect(DatefieldComp).toBeInTheDocument();
  });

  it("Should render with an end date", async () => {
    const { container } = render(
      <DateRange
        data-testid="DateRange"
        startInputId="test-6"
        endInputId="test-6"
        endInputName="test-6"
        startInputName="test-6"
        startLabel="test-6"
        endLabel="test-6"
        endDate="09/10/2022"
        hint={true}
      />
    );
    const DatefieldButton = container.getElementsByClassName(
      "usa-date-picker__button"
    )[0];
    fireEvent.click(DatefieldButton);
    await new Promise((res) => setTimeout(res, 1000));

    // expect(endDate).toHaveFocus();
  });
});

describe("compontent snapshots", () => {
  it("primary datefield", () => {
    const { container } = render(
      <DateRange
        hint
        startInputId="test-3"
        endInputId="test-3"
        startInputName="test-3"
        endInputName="test-3"
        startLabel="test-3"
        endLabel="test-3"
      />
    );

    expect(container).toMatchSnapshot();
  });

  it("max date datefield", () => {
    const { container } = render(
      <DateRange
        hint
        startInputId="test-4"
        startInputName="test-4"
        startLabel="test-4"
        endInputId="test-4"
        endInputName="test-4"
        endLabel="test-4"
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("min date datefield", () => {
    const { container } = render(
      <DateRange
        hint={false}
        startInputId="test-5"
        startInputName="test-5"
        startLabel="test-5"
        endInputId="test-5"
        endInputName="test-5"
        endLabel="test-5"
      />
    );
    expect(container).toMatchSnapshot();
  });
});
