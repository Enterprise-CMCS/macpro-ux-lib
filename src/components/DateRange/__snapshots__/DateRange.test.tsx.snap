// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compontent snapshots max date datefield 1`] = `
<div>
  <div
    class="usa-date-range-picker"
  >
    <div
      class="usa-form-group datefield"
    >
      <label
        class="usa-label"
        for="test-4"
        id="test-4-label"
      >
        test-4
      </label>
      <div
        class="usa-hint"
        id="test-4-hint"
      >
        mm/dd/yyyy
      </div>
      <div
        class="usa-date-picker"
        data-max-date="false"
        data-min-date="false"
      >
        <input
          aria-describedby="test-4-hint"
          aria-labelledby="test-4-label"
          class="usa-input"
          id="test-4"
          name="test-4"
        />
      </div>
    </div>
    <div
      class="usa-form-group datefield"
    >
      <label
        class="usa-label"
        for="test-4"
        id="test-4-label"
      >
        test-4
      </label>
      <div
        class="usa-hint"
        id="test-4-hint"
      >
        mm/dd/yyyy
      </div>
      <div
        class="usa-date-picker"
        data-max-date="false"
        data-min-date="false"
      >
        <input
          aria-describedby="test-4-hint"
          aria-labelledby="test-4-label"
          class="usa-input"
          id="test-4"
          name="test-4"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`compontent snapshots min date datefield 1`] = `
<div>
  <div
    class="usa-date-range-picker"
  >
    <div
      class="usa-form-group datefield"
    >
      <label
        class="usa-label"
        for="test-5"
        id="test-5-label"
      >
        test-5
      </label>
      <div
        class="usa-date-picker"
        data-max-date="false"
        data-min-date="false"
      >
        <input
          aria-describedby="test-5-label"
          aria-labelledby="test-5-label"
          class="usa-input"
          id="test-5"
          name="test-5"
        />
      </div>
    </div>
    <div
      class="usa-form-group datefield"
    >
      <label
        class="usa-label"
        for="test-5"
        id="test-5-label"
      >
        test-5
      </label>
      <div
        class="usa-date-picker"
        data-max-date="false"
        data-min-date="false"
      >
        <input
          aria-describedby="test-5-label"
          aria-labelledby="test-5-label"
          class="usa-input"
          id="test-5"
          name="test-5"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`compontent snapshots primary datefield 1`] = `
<div>
  <div
    class="usa-date-range-picker"
  >
    <div
      class="usa-form-group datefield"
    >
      <label
        class="usa-label"
        for="test-3"
        id="test-3-label"
      >
        test-3
      </label>
      <div
        class="usa-hint"
        id="test-3-hint"
      >
        mm/dd/yyyy
      </div>
      <div
        class="usa-date-picker"
        data-max-date="false"
        data-min-date="false"
      >
        <input
          aria-describedby="test-3-hint"
          aria-labelledby="test-3-label"
          class="usa-input"
          id="test-3"
          name="test-3"
        />
      </div>
    </div>
    <div
      class="usa-form-group datefield"
    >
      <label
        class="usa-label"
        for="test-3"
        id="test-3-label"
      >
        test-3
      </label>
      <div
        class="usa-hint"
        id="test-3-hint"
      >
        mm/dd/yyyy
      </div>
      <div
        class="usa-date-picker"
        data-max-date="false"
        data-min-date="false"
      >
        <input
          aria-describedby="test-3-hint"
          aria-labelledby="test-3-label"
          class="usa-input"
          id="test-3"
          name="test-3"
        />
      </div>
    </div>
  </div>
</div>
`;
